// core/main-unittest.js
var
    options = require('options'),
    mainSetup = false,
    allTests = {},
    testCount = 0,
    server = {},
    schemas = [];

options.platform_port = options.platform_port + 1;

var requireMain = function (callback) {
    var
        cback = function (xsrv) {
            server = xsrv.server;
            allTests.groupN = {
                testO: function (test) {
                    test.done();
                }
            };
            xsrv.schemas.forEach(function (schema) {
                allTests[schema] = {
                    testX: function (test) {
                        test.done();
                    }
                };
            });
            console.log(allTests);


            callback();
        };
    if (!mainSetup) {
        console.log(allTests);
        for (var i in allTests) {
            if (/test.*/.test(i))
                testCount = testCount + 1;
        }
        require('core/main')(options, cback);
        mainSetup = true;
    } else
        callback();
};

allTests = {
    setUp: function (callback) {
        requireMain(callback);
    },
    groupM: {
        testN: function (test) {
            test.done();
        }
    },
    testSomething: function (test) {
        test.ok('Hello');
        test.done();
    },
    testOlolo: function (test) {
        test.equals(1, 1);
        test.done();
    },
    tearDown: function (callback) {
        testCount = testCount - 1;
        server.on('close', function () {
            setTimeout(function () {
                process.nextTick(function () {
                    process.exit();
                })
            }, 1000);
        });
        if (testCount == 0)
            server.close();
        callback();
    }
};

module.exports = allTests;
