// core/db/mongodb/schema/model.js
var
    access = require('core/access.js'),
    db = require('mongoose'),
    Schema = db.Schema,
    ObjectId = Schema.ObjectId;

/**
 * DBSchema describing schemas
 *
 *      - name - name of the schema
 *      - level:
 *          - 0 - core schema, using by clients is prohibited
 *          - 1 - app schema, using by Ajax clients is prohibited
 *          - 2 - same, but allowed if Ajax clients have been authorized
 *          - 3 - schema may be accessed by everyone
 *          - 4 - reserved
 *
 */

var schemaDescription = {
    LEVEL_CORE: 0,
    LEVEL_APPS: 1,
    LEVEL_AUTH: 2,
    LEVEL_AJAX: 3,
    LEVEL_RESV: 4
};

var schemaModel;

schemaDescription.schemaSchema = new Schema({
    id: ObjectId,
    name: { type: String },
    level: { type: Number,
        min: access.LEVEL_CORE, max: access.LEVEL_RESV
    }
});

db.model('schemaModel', schemaDescription.schemaSchema);
schemaModel = db.model('schemaModel');

module.exports = exports = {
    description: schemaDescription,
    model: schemaModel
};
